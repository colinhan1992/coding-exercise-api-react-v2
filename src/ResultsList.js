import React, { Fragment, Component } from 'react';
import { Table } from 'semantic-ui-react';
import PropTypes from 'prop-types';

import GroupMembers from './GroupMembers';

class ResultsList extends Component {
    static propTypes = {
        groups: PropTypes.array.isRequired,
        people: PropTypes.array.isRequired
    };

    getGroupName = groupID => {
        const group = this.props.groups.find(e => e.id === groupID);
        return group ? group.group_name : 'N/A';
    };

    getGroupMembers(groupID) {
        const members = this.props.people.filter(e => e.group_id === groupID);
        return members.length;
    }

    render() {
        const { groups, people } = this.props;

        const groupsWithMembers = groups.filter(function(group) {
            return people.find(p => p.group_id === group.id) !== undefined;
        });

        return (
            <Fragment>
                <h3>Groups</h3>
                <Table sortable celled padded>
                    <Table.Header>
                        <Table.Row>
                            <Table.HeaderCell>Group Name</Table.HeaderCell>
                            <Table.HeaderCell>Members</Table.HeaderCell>
                        </Table.Row>
                    </Table.Header>

                    <Table.Body>
                        {groups.map((group, index) => {
                            return (
                                <Table.Row key={index}>
                                    <Table.Cell singleLine>
                                        {group.group_name}
                                    </Table.Cell>
                                    <Table.Cell singleLine>
                                        {this.getGroupMembers(group.id)}
                                    </Table.Cell>
                                </Table.Row>
                            );
                        })}
                    </Table.Body>
                </Table>
                {groupsWithMembers.map((group, index) => {
                    return (
                        <GroupMembers
                            key={index}
                            group={group}
                            members={people.filter(
                                p => p.group_id === group.id
                            )}
                            getGroupName={this.getGroupName}
                        />
                    );
                })}
            </Fragment>
        );
    }
}

export default ResultsList;
