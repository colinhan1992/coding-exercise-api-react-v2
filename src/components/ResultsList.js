import React, { Fragment, Component } from 'react';
import { Table } from 'semantic-ui-react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import GroupMembers from './GroupMembers';

class ResultsList extends Component {
    static propTypes = {
        groups: PropTypes.array,
        people: PropTypes.array
    };

    state = {
        column: null,
        data: this.props.groups.map(group => {
            return {
                id: group.id,
                group_name: group.group_name,
                members: this.getGroupMembers(group.id)
            };
        }),
        direction: null
    };

    handleSort = clickedColumn => () => {
        const { column, data, direction } = this.state;

        if (column !== clickedColumn) {
            this.setState({
                column: clickedColumn,
                data: _.sortBy(data, [clickedColumn]),
                direction: 'ascending'
            });

            return;
        }

        this.setState({
            data: data.reverse(),
            direction: direction === 'ascending' ? 'descending' : 'ascending'
        });
    };

    getGroupName = groupID => {
        const group = this.props.groups.find(e => e.id === groupID);
        return group ? group.group_name : 'N/A';
    };

    getGroupMembers(groupID) {
        const members = this.props.people.filter(e => e.group_id === groupID);
        return members.length;
    }

    render() {
        const { groups, people } = this.props;
        const { data, column, direction } = this.state;

        const groupsWithMembers = groups.filter(function(group) {
            return people.find(p => p.group_id === group.id) !== undefined;
        });

        return (
            <Fragment>
                <h3>Groups</h3>
                <Table sortable celled padded>
                    <Table.Header>
                        <Table.Row>
                            <Table.HeaderCell
                                sorted={
                                    column === 'group_name' ? direction : null
                                }
                                onClick={this.handleSort('group_name')}
                            >
                                Group Name
                            </Table.HeaderCell>
                            <Table.HeaderCell
                                sorted={column === 'members' ? direction : null}
                                onClick={this.handleSort('members')}
                            >
                                Members
                            </Table.HeaderCell>
                        </Table.Row>
                    </Table.Header>

                    <Table.Body>
                        {data.map((group, index) => {
                            return (
                                <Table.Row key={index}>
                                    <Table.Cell singleLine>
                                        {group.group_name}
                                    </Table.Cell>
                                    <Table.Cell singleLine>
                                        {group.members}
                                    </Table.Cell>
                                </Table.Row>
                            );
                        })}
                    </Table.Body>
                </Table>
                {groupsWithMembers.map((group, index) => {
                    return (
                        <GroupMembers
                            key={index}
                            group={group}
                            members={people.filter(
                                p => p.group_id === group.id
                            )}
                        />
                    );
                })}
            </Fragment>
        );
    }
}

export default ResultsList;
