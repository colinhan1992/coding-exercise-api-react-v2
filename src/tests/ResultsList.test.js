import React from 'react';
import { shallow } from 'enzyme';
import ResultsList from '../components/ResultsList';

let wrapper, data;

describe('<ResultsList />', () => {
    beforeAll(() => {
        data = [
            {
                id: 132,
                group_name: 'Volunteers',
                updated_at: '2019-07-20 22:05:47',
                created_at: '2019-07-20 22:05:47'
            }
        ];
        wrapper = shallow(<ResultsList people={[]} groups={data} />);

        wrapper.setState({ data: data });
    });

    test('should match the snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
});
