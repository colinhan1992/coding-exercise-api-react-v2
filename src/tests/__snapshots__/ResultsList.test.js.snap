// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultsList /> should match the snapshot 1`] = `
<Fragment>
  <h3>
    Groups
  </h3>
  <Table
    as="table"
    celled={true}
    padded={true}
    sortable={true}
  >
    <TableHeader
      as="thead"
    >
      <TableRow
        as="tr"
        cellAs="td"
      >
        <TableHeaderCell
          as="th"
          onClick={[Function]}
          sorted={null}
        >
          Group Name
        </TableHeaderCell>
        <TableHeaderCell
          as="th"
          onClick={[Function]}
          sorted={null}
        >
          Active Members
        </TableHeaderCell>
      </TableRow>
    </TableHeader>
    <TableBody
      as="tbody"
    >
      <TableRow
        as="tr"
        cellAs="td"
        key="0"
      >
        <TableCell
          as="td"
          singleLine={true}
        >
          Volunteers
        </TableCell>
        <TableCell
          as="td"
          singleLine={true}
        />
      </TableRow>
    </TableBody>
  </Table>
</Fragment>
`;
